### PR TITLE
[Enhancement] Refine profile merge mechanism

### LINKED ISSUE #9356: [Enhancement] Make the meaning of profile metric min/max stable
## Background

In pipeline engine, originl profile has 5 levels, fragment, fragment instance, pipeline, pipeline driver, operator, respectively.

image

By default(`pipeline_profile_level=1`), some isomorphic parts will be merged out, for example, pipeline driver level will be merged in to pipeline, fragment instance level will be merged into fragment. 

So, 3 levels remains after profile merging, fragment, pipeline, operator, respectively.

image

When performing merging, metric will be merged according to its type
* time metric, using average as merged value.
* non-time metric, using grand-total sum as merged value. 

## Problem

And here is the problem, we only add `__MIN_xxx` or `__MAX_xxx` of metric `xxx` if it varies greatly. 

But for non-time metric, the meaning of `__MIN_xxx` or `__MAX_xxx` are different

* pipeline-driver level merging, `__MIN_xxx` or `__MAX_xxx` means the min/max value of among all the pipeline_driver's
* fragment-instance level merging, `__MIN_xxx` or `__MAX_xxx` means the min/max value among pipeline's, which is the already the sum of all the pipeline driver level's

So after the two level merging, `__MIN_xxx` means the min value among all pipeline_driver's while `__MAX_xxx` means the max value among all pipeline's. It is confused!

For example

```
# first merge
(1,2,3) -> 6[MIN=1, MAX=3]
(4,5,6) -> 15[MIN=4, MAX=6]
(7,8,9) -> 24[MIN=7, MAX=9]

# second merge
(6[MIN=1, MAX=3], 15[MIN=4, MAX=6)], 24[MIN=7, MAX=9]) -> 45[MIN=1, MAX=24]
```

After the two stage merge, we get the `45[MIN=1, MAX=24]`, 1 is the minimum value among all isomorphic pipeline_driver's while 24 is the maximum value among all isomorphic pipeline's. We expected to get the final metric as `45[MIN=1, MAX=9]`

### PR DESCRIPTION
**Pull request by liuyehcf**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #9356

## Problem Summary(Required) ：

Refer to #9356 for background of the problem.

## Solution

For all non-time metrics, we calculate `__MIN_xxx`、`__MAX_xxx`, and DO NOT perform the second merge if current metric is already a merged one, just pick the minimum one among `__MIN_xxx` and the maximum one among `__MAX_xxx` as the new `__MIN_xxx`、`__MAX_xxx` of current metric.

For example

**Before**

```
# first merge
(1,2,3) -> 6[MIN=1, MAX=3]
(4,5,6) -> 15[MIN=4, MAX=6]
(7,8,9) -> 24[MIN=7, MAX=9]

# second merge
(6[MIN=1, MAX=3], 15[MIN=4, MAX=6)], 24[MIN=7, MAX=9]) -> 45[MIN=1, MAX=24]
```

**After**

```
# first merge
(1,2,3) -> 6[MIN=1, MAX=3]
(4,5,6) -> 15[MIN=4, MAX=6]
(7,8,9) -> 24[MIN=7, MAX=9]

# second merge
(6[MIN=1, MAX=3], 15[MIN=4, MAX=6)], 24[MIN=7, MAX=9]) -> 45[MIN=1, MAX=9]
```

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 37 / 37 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/RuntimeProfile.java|36|36|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/Counter.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 31 / 31 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/RuntimeProfile.java|30|30|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/Counter.java|1|1|100.00%|[]|

File: be/src/util/runtime_profile.cpp [hunk 1]
```diff
@@ -1030,7 +1030,11 @@ void RuntimeProfile::merge_isomorphic_profiles(std::vector<RuntimeProfile*>& pro
                 continue;
             }
 
-            std::vector<std::tuple<Counter*, Counter*, Counter*>> counters;
+            std::vector<Counter*> counters;
+
+            int64_t min_value = std::numeric_limits<int64_t>::max();
+            int64_t max_value = std::numeric_limits<int64_t>::min();
+            bool already_merged = false;
```
Comment (mofeiatwork) [inline be/src/util/runtime_profile.cpp:1037]: Does it need two flag to indicates min and max merged?

**Review by murphyatwork — DISMISSED** [review]

**Comment by ZiheLiu** [comment]
run starrocks_admit_test

**Review by ZiheLiu — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by ZiheLiu — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 31 / 31 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/RuntimeProfile.java|30|30|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/Counter.java|1|1|100.00%|[]|

**Review by murphyatwork — DISMISSED** [review]

**Comment by liuyehcf** [comment]
run  starrocks_admit_test

**Comment by liuyehcf** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/common/util/RuntimeProfile.java [hunk 2]
```diff
@@ -428,7 +427,7 @@ public String getInfoString(String key) {
         return infoStrings.get(key);
     }
 
-    // Merge all the isomorphic sub profiles and the caller must known for sure
+    // Merge all the isomorphic sub profiles and the caller must know for sure
     // that all the children are isomorphic, otherwise, the behavior is undefined
     // The merged result will be stored in the first profile
     public static void mergeIsomorphicProfiles(List<RuntimeProfile> profiles) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/common/util/RuntimeProfile.java:433]: The FE `mergeIsomorphicProfiles` implementation is same with `merge_isomorphic_profiles`?

Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/common/util/RuntimeProfile.java:433]: Yes, exactly the same.

**Review by kangkaisen — DISMISSED** [review]

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 31 / 31 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/RuntimeProfile.java|30|30|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/Counter.java|1|1|100.00%|[]|

**Comment by liuyehcf** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 31 / 31 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/RuntimeProfile.java|30|30|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/Counter.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 49 / 51 (96.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/RuntimeProfile.java|48|50|96.00%|[558, 564]|
|:large_blue_circle:|com/starrocks/common/util/Counter.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 34 / 34 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/RuntimeProfile.java|33|33|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/Counter.java|1|1|100.00%|[]|

**Review by kangkaisen — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Pull request merged**